sdk/iot/Azure.Iot.Hub.Service/src/API Design.md

```diff
@@ -42,7 +42,129 @@ APIs for getting statistics about devices and modules, as well as service statis
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>The result of the bulk operation.</returns>
+    public virtual async Task<Response<BulkRegistryOperationResult>> CreateIdentitiesAsync(IEnumerable<DeviceIdentity> deviceIdentities, CancellationToken cancellationToken = default)
+
+    /// <summary>
+    /// Update multiple devices. A maximum of 100 updates can be done per call, and each operation must be done on a different identity. For larger scale operations, consider using IoT Hub jobs (https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-identity-registry#import-and-export-device-identities).
+    /// </summary>
+    /// <param name="deviceIdentities">The devices to update.</param>
+    /// <param name="force">If true, the devices will be updated even if their ETag is out of date.
+    /// If false, each device will only be updated if its ETag is up to date.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>The result of the bulk operation.</returns>
+    public virtual async Task<Response<BulkRegistryOperationResult>> UpdateIdentiesAsync(IEnumerable<DeviceIdentity> deviceIdentities, bool force, CancellationToken cancellationToken = default)
+    
+    /// <summary>
+    /// Delete multiple devices. A maximum of 100 deletions can be done per call. For larger scale operations, consider using IoT Hub jobs (https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-identity-registry#import-and-export-device-identities).
+    /// </summary>
+    /// <param name="deviceIdentities">The devices to delete.</param>
+    /// <param name="force">If true, the devices will be deleted even if their ETag is out of date.
+    /// If false, each device will only be deleted if its ETag is up to date.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>The result of the bulk deletion.</returns>
+    public virtual async Task<Response<BulkRegistryOperationResult>> DeleteIdentitiesAsync(IEnumerable<DeviceIdentity> deviceIdentities, bool force, CancellationToken cancellationToken = default)
+
+    /// <summary>
+    /// List a set of device twins.
+    /// </summary>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>A pageable set of device twins.</returns>
+    public virtual AsyncPageable<TwinData> GetTwinsAsync(CancellationToken cancellationToken = default)
+
+    /// <summary>
+    /// Get a device's twin.
+    /// </summary>
+    /// <param name="deviceId">The unique identifier of the device to get the twin of.</param>
+    /// <param name="cancellationToken">The cancellation token</param>
+    /// <returns>The device's twin, including reported properties and desired properties.</returns>
+    public virtual async Task<Response<TwinData>> GetTwinAsync(string deviceId, CancellationToken cancellationToken = default)
+
```

Comment:
The only concern I have with this design is follows : 
1. I have to maintain mappings of device identities to twin data data types for a combination of twin/identity operations
2. When it comes to twin/methods the object "Device Identity" that I stored carries no value except that I had to store it entirely to retrieve device ID.
3. Using "force" isn't clear to me that it is replace. does it warrant a separate API ?

Reply:
1.) Yes, this is a bothersome design flaw. But this is a design flaw at the service level, not the SDK level. Did you have something in mind to fix it at the SDK level?
2.) Correct. Most device identity fields are only useful in the context of device identity management, not twin or methods. But that's okay, because you can update device identity with that "device identity" that you stored
3.) Force implies to ignore etag comparisons when updating or deleting. I outline that in the parameter description pretty thoroughly, so I'm not sure what you want me to add here.

Reply:
1 ) Can we use CL to clarify this for our customers and propose relevant changes to iothub eventually? Perhaps a single uber object which is a combination of twin data + device identity. Starting at least from CL and eventually suggested to iothub team to fix. This will fix this particular design flaw.
2 ) 1 will fix this problem as well.
3) I am open to reducing API's count and starting with this.